src/vs/platform/quickinput/browser/quickPickPin.ts

```diff
@@ -0,0 +1,107 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Codicon } from 'vs/base/common/codicons';
+import { localize } from 'vs/nls';
+import { ServicesAccessor } from 'vs/platform/instantiation/common/instantiation';
+import { IQuickPick, IQuickPickItem, QuickPickItem } from 'vs/platform/quickinput/common/quickInput';
+import { IStorageService, StorageScope, StorageTarget } from 'vs/platform/storage/common/storage';
+import { ThemeIcon } from 'vs/platform/theme/common/themeService';
+
+const pinButtonClass = ThemeIcon.asClassName(Codicon.pin);
+const pinnedButtonClass = ThemeIcon.asClassName(Codicon.pinned);
+const buttonClasses = [pinButtonClass, pinnedButtonClass];
+/**
+ * Initially, adds pin buttons to all @param quickPick items.
+ * When pinned, a copy of the item will be moved to the end of the pinned list and any duplicate within the pinned list will
+ * be removed if @param filterDupliates has been provided. Pin and pinned button events trigger updates to the underlying storage.
+ * Shows the quickpick once formatted.
+ */
+export async function showWithPinnedItems(accessor: ServicesAccessor, storageKey: string, quickPick: IQuickPick<IQuickPickItem>, filterDuplicates?: boolean): Promise<void> {
+	const storageService = accessor.get(IStorageService);
```

Comment:
So by using `ServicesAccessor` you are forcing everyone to include `InstantiationService` in their dependency injection.

You could equally just have them pass in the `IStorageService` and then they don't need to include `InstantiationService` just `IStorageService`.

Another benefit is that you can call this function normally:
```
showWithPinnedItems(this._storageService, runRecentStorageKey, quickPick, true)
```
instead of using 
```
this._instantiationService.invokeFunction(showWithPinnedItems, runRecentStorageKey, quickPick, true);
```

Reply:
I know @Tyriar prefers the instantiation service invoke function.... I think it's more likely that everyone will have that than they will the storageService

Reply:
I don't think this is the sort of thing that should prevent this PR from being merged. once it is, since it's your area, feel free to change 😄 would be good to merge so I can do more testing before endgame

Reply:
I prefer the first way too, so am reverting to that